code/components/gta-streaming-five/src/GameEventHooks.cpp

```diff
@@ -62,18 +107,144 @@ void* HandleEventWrap(void* group, rage::fwEvent* event)
+	{
+		/*
+		 * event:
+		 * These events are stored with different member variables and with different layouts per class/struct,
+		 * so we have to investigate them case by case.
+		 * 
+		 * group:
+		 * In this reacting version the entity that triggers this event is stored in group->entities[17]
+		 * as well as in group->entities[18].
+		 */
 
-namespace rage
+		try
+		{
+			msgpack::sbuffer buf;
+			msgpack::packer packer(buf);
+
+			const char* eventName = typeid(*event).name() + 6;
+
+			packer.pack_array(4); // we'll offer 4 parameters
+			msgpack_pack(packer, eventName, event->GetId(), GetGuidFromBaseSafe(group->entities[17]));
+
+			// Quick 4 bytes check, i.e.: no strcmp
+			// this will turn the first 4 bytes after "CEvent" into an uint32 and check against that value
+			switch ((uint32_t&)eventName[6])
+			{
+				case ct_uint32("Shoc"): // Shocking
+				{
+					msgpack_array(packer, GetGuidFromBaseSafe((rage::fwEntity*&)event[13]), (float(&)[3])event[8]);
+					break;
+				}
+				case ct_uint32("Pote"): // Potential
+				{
+					// TODO: check how we will do this without a strcmp
+					if (strcmp(eventName + 15, "WalkIntoVehicle") == 0)
+					{
+						msgpack_array(packer, GetGuidFromBaseSafe((rage::fwEntity*&)event[8]), (float(&)[3])event[5]);
+					}
+					else if (strcmp(eventName + 15, "GetRunOver") == 0)
+					{
+						msgpack_array(packer, GetGuidFromBaseSafe((rage::fwEntity*&)(event[6])));
+						// doesn't seem there is a vector3, might have other info, speed, vector2?
+					}
+					else
+					{
+						packer.pack_array(0); // empty array
+					}
+					break;
+				}
+				default:
+				{
+					packer.pack_array(0); // empty array
+					break;
+				}
+			}
+
+			OnTriggerGameEventReact({ buf.data(), buf.size() });
+		}
+		catch (std::exception& e)
+		{
+		}
+	}
+
+	return (*TFunc)(group, event);
+}
+
+template<decltype(g_eventCall2)* TFunc>
+void* HandleEventWrapEmit(rage::fwEventGroup* group, rage::fwEvent* event)
 {
-class netObject;
+	if (event)
+	{
+		/*
+		 * event:
+		 * These events are stored with different member variables and with different layouts per class/struct,
+		 * so we have to investigate them case by case.
+		 * 
+		 * group:
+		 * In this emitting version they store them from group->entities[0] to group->entities[n] 
+		 * where it ends with 0x0 or a repeating entity ptr.
+		 */
+
+		try
+		{
+			msgpack::sbuffer buf;
+			msgpack::packer packer(buf);
+
+			const char* eventName = typeid(*event).name() + 6;
+
+			packer.pack_array(4); // we'll offer 4 parameters
+			msgpack_pack(packer, eventName, event->GetId());
+
+			// retrieve all entities
+			rage::fwEntity* ent2 = nullptr; // to ignore duplicates on the end
+			rage::fwEntity** ent1 = group->entities; // to ignore duplicates on the end
+			
+			// iterate ptrs until we find a nullptr, a duplicate, or if we reached the end of the buffer size
+			size_t size = 0;
+			for (; *ent1 && *ent1 != ent2 && ++size < _countof(group->entities); ent2 = *ent1++);
```

Comment:
Cute, but bad for readability.

Reply:
True that's why I commented it.

Reply:
On this note, I could also just fill the array and change the size bytes later of the serialization. Though I would think that's generally frowned upon, what's your opinion?